make TensorImpl::data_ptr_impl() non-const and have mutable in the name

dagitses opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98162
* #98163
* #97874
* #97804
* #97776
* __->__ #97744

See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

Commit:
make TensorImpl::data_ptr_impl() non-const and have mutable in the name

See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

Commit:
Update on "make TensorImpl::data_ptr_impl() non-const and have mutable in the name"


See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

Commit:
Update on "make TensorImpl::data_ptr_impl() non-const and have mutable in the name"


See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

Commit:
Update on "make TensorImpl::data_ptr_impl() non-const and have mutable in the name"


See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

Commit:
Update on "make TensorImpl::data_ptr_impl() non-const and have mutable in the name"


See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

Commit:
Update on "make TensorImpl::data_ptr_impl() non-const and have mutable in the name"


See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

Commit:
Update on "make TensorImpl::data_ptr_impl() non-const and have mutable in the name"


See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/templates/TensorMethods.cpp:
```diff
@@ -14,7 +14,7 @@ namespace at {
          " but found ",                                              \
          scalar_type());                                             \
-     return this->unsafeGetTensorImpl()->data_ptr_impl<T>();         \
+     return this->unsafeGetTensorImpl()->mutable_data_ptr_impl<T>(); \
```
Is this temporary, is that right?

dagitses replied:
Correct. If this eventually is intended to return a const pointer, it will not call the mutable variant.

Commit:
Update on "make TensorImpl::data_ptr_impl() non-const and have mutable in the name"


See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

Commit:
Update on "make TensorImpl::data_ptr_impl() non-const and have mutable in the name"


See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

Commit:
Update on "make TensorImpl::data_ptr_impl() non-const and have mutable in the name"


See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

Commit:
Update on "make TensorImpl::data_ptr_impl() non-const and have mutable in the name"


See D44409928.

Differential Revision: [D44450468](https://our.internmc.facebook.com/intern/diff/D44450468/)

[ghstack-poisoned]

t10-13rocket reviewed (commented):
LGTM
